### PR TITLE
Fix unexpected exception return behavior

### PR DESCRIPTION
**Pull request by yenWu**
Just like what I confuse in #14 
I write additional code to check the situation and omit unnecessary condition switching to unprivileged thread mode.

### CODE REVIEW

File: 04-Multitasking/context_switch.S [hunk 1]
```diff
@@ -1,5 +1,16 @@
 .syntax unified
 
+/* check the situaion and determine the path to return */
+.macro ret_path, scratch
+	mov \scratch, #0xfffffff0
+	cmp lr, \scratch  /* lr - 0xfffffff0 */
+
+	ittt ls /* `ls` condition means unsinged lower or the same */
```
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Avoid using C-style comment. Instead, use `@`.

Comment (jserv) [inline 04-Multitasking/context_switch.S]: How many times does macro `ret_path` have been used so far?

Comment (yenWu) [inline 04-Multitasking/context_switch.S]: Only once. But I think it is more clear and flexible than leave hard code in `activate()`.

Comment (yenWu) [inline 04-Multitasking/context_switch.S]: @jserv , The original coding style in `context_switch.S` use C-style comment. Should I modify all of them?

Comment (jserv) [inline 04-Multitasking/context_switch.S]: Use `@` for inlined comments. Use C-style comments outside assembly block.

**Comment by jserv** [comment]
Rebase first in order to eliminate the duplicated commit, "Fix the deadlinks and typo".

File: 04-Multitasking/context_switch.S (hunk 1 above)
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Macro `ret_path` is not meant to be useful to common cases. I would expect replacing it back to `activate`.

**Comment by yenWu** [comment]
@jserv , Done! Thank for your useful comments.

File: 04-Multitasking/context_switch.S [hunk 2]
```diff
@@ -19,14 +19,32 @@ activate:
...
+	ldmia r0!, {r4, r5, r6, r7, r8, r9, r10, r11, lr}
+
+	msr psp, r0
+
+	/* check the situaion and determine the path to return */
+	mov r0, #0xfffffff0
+	cmp lr, r0                      @ lr - 0xfffffff0
+
+	/* switch to process stack */
+	ittt ls                         @ `ls` condition means unsinged lower or the same
```
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Use quote (") to mark "ls".

File: 04-Multitasking/context_switch.S [hunk 3]
```diff
@@ -19,14 +19,32 @@ activate:
...
+	movls r0, #3
+	msrls control, r0
+	isbls
+
+	/* jump to user task */
+	bx lr/* check the situaion and determine the path to return */
```
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Change comment style.

File: 04-Multitasking/context_switch.S [hunk 4]
```diff
@@ -19,14 +19,32 @@ activate:
...
+	mov r0, #0xfffffff0
+	cmp lr, r0                      @ lr - 0xfffffff0
+
+	/* switch to process stack */
+	ittt ls                         @ `ls` condition means unsinged lower or the same
```
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Ditto.

File: 04-Multitasking/context_switch.S [hunk 5]
```diff
@@ -19,14 +19,32 @@ activate:
...
+	movls r0, #3
+	msrls control, r0
+	isbls
+
+	/* jump to user task */
+	bx lr/* check the situaion and determine the path to return */
+	mov r0, #0xfffffff0
+	cmp lr, r0                      @ lr - 0xfffffff0
+
+	/* switch to process stack */
```
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Why do you have to switch to process stack twice?

File: 04-Multitasking/context_switch.S (hunk 4 above)
Comment (yenWu) [inline 04-Multitasking/context_switch.S]: I can understand the previous one, but what's wrong with this section?  Should I change "/* switch to process stack */" to "@switch to process stack" ? I just follow the original coding-style.

File: 04-Multitasking/context_switch.S (hunk 5 above)
Comment (yenWu) [inline 04-Multitasking/context_switch.S]: I would improve semantic of comments.

File: 04-Multitasking/context_switch.S (hunk 4 above)
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Clarify: ditto for Line 39.

File: 04-Multitasking/context_switch.S [hunk 6]
```diff
@@ -19,14 +19,32 @@ activate:
...
-	msr control, r0
-	isb
 
 	/* load user state */
-	pop {r4, r5, r6, r7, r8, r9, r10, r11, lr}
+	ldmia r0!, {r4, r5, r6, r7, r8, r9, r10, r11, lr}
+
+	msr psp, r0
+
+	/* check the situaion and determine the path to return */
```
Comment (jserv) [inline 04-Multitasking/context_switch.S]: Why would you mention it as the path to return? I don't really think it is a path. Instead, it is transition.

File: 04-Multitasking/context_switch.S (hunk 5 above)
Comment (yenWu) [inline 04-Multitasking/context_switch.S]: I took a mistake on it with duplicated code.

**Comment by yenWu** [comment]
@jserv , Fix inappropriate semantic and coding-style

**Comment by yenWu** [comment]
@jserv , I already rebase to upstream.

File: 04-Multitasking/context_switch.S [hunk 7]
```diff
@@ -19,14 +19,21 @@ activate:
...
-	pop {r4, r5, r6, r7, r8, r9, r10, r11, lr}
+	ldmia r0!, {r4, r5, r6, r7, r8, r9, r10, r11, lr}
+
+	msr psp, r0
+
+	/* check the situaion and determine the transition */
+	mov r0, #0xfffffff0
+	cmp lr, r0                      @ lr - 0xfffffff0
+
+	/* if "lr" is not exception return, then switch to process stack */
```
Comment (jserv) [inline 04-Multitasking/context_switch.S]: The statement should "if lr does not point to exception return...".

**Comment by jserv** [comment]
Rather than "I add a few lines in `activate()`", you can say, "Change the operation of activate() to ...".

**Comment by yenWu** [comment]
@jserv , Done! Thank for your patience. By the way, I also rebase to upstream.

**Comment by jserv** [comment]
There is a typo in subject of Git commit message. Should be "exception".

**Pull request merged**